### ISSUE TITLE
rustdoc "Aliased Type" view is wrong

### ISSUE THREAD

**Issue by SludgePhD** [rust-lang/rust#120404]
![screenshot_2024-01-27_01:56:48](https://github.com/rust-lang/rust/assets/96552222/b659d0ab-1901-4fc8-8c5f-855767fe6008)

In this rustdoc output, rustdoc seems to claim that `MatrixView` aliases a `struct MatrixView`.

This is wrong, since no such `struct` exists. It seems like there was some sort of mistake when implementing this feature that ended up using the wrong name there? It should say `Matrix` instead.

**Comment by SludgePhD** [comment]
*Any* type alias of a `struct` will reproduce it.

```rust
pub type Alias<T> = Vec<T>;
```

Result:

![screenshot_2024-01-27_02:54:58](https://github.com/rust-lang/rust/assets/96552222/d8d180d8-e9fe-4e79-8ea7-5cb3e5072bf5)

**Comment by fmease** [comment]
This works exactly as intended: The “aliased type” shown isn't a real user-written type but a synthetic one.

**Comment by fmease** [comment]
Of course, we should probably employ some more heuristics and e.g., not show aliased structs whose fields are all private.

**Comment by fmease** [comment]
I can understand that that feature might be a bit unintuitive at first. If you have suggestions for making it clearer to the reader that this aliases type isn't grounded in source code, please don't hesitate to share them :)

**Comment by SludgePhD** [comment]
To me, "aliased type" strictly refers to the right-hand side of the `type A = B` syntax, so I would find this feature much less confusing if it simply used the name of the right-hand type instead.

**Comment by Urgau** [comment]
> To me, "aliased type" strictly refers to the right-hand side of the `type A = B` syntax, so I would find this feature much less confusing if it simply used the name of the right-hand type instead.

@SludgePhD Except that this wouldn't work in "generic context", take for exemple this type alias:
```rust
type NonZeroU128 = NonZero<u128>;
```
writting it with the "right hand side" would mean:
```rust
struct NonZero<u128>(/* ... */);
```
which is not a valid syntax and is not that what anyone that would use `NonZeroU128` would need to know.

Taking a more expressive example:
```rust
struct Container<A: Trait>(pub A, pub A);
type MyContainer = Container<MyStruct, MyStruct>;
```

Here, the aliased type (as seen by the compiler) is:
```rust
struct MyContainer(pub MyStruct, pub MyStruct);
```
and it makes much more sense to write it like this. That way users don't have the irrelevant `Trait` bound or even the generic in the way.

------

You can look at the description of https://github.com/rust-lang/rust/pull/114855 for a more detailed explanation and motivation for the "aliased type" feature.

------

As for how we can improve the reading of this section, what about a tooltip (or small label) with something like this:
> This section shows the concrete type (i.e. removing any generics/bounds not present the aliased type) has would been seen that anyone using the aliased type.

**Comment by lolbinarycat** [comment]
I think the logical change here would be to change the "Aliased Type" heading to something like "Effective Type" or something.  This wouldn't be the first time rustdoc has made significant wording changes (see the whole "Trait Object") thing.

**Comment by Urgau** [comment]
> This wouldn't be the first time rustdoc has made significant wording changes (see the whole "Trait Object") thing.

Unless I'm mistaken that change was initiated by T-lang, rustdoc just followed the move.

"Aliased Type" on the other hand is a term I made up, so feel free to change it so something clearer.

**Comment by lolbinarycat** [comment]
Just created [an IRLO thread](https://internals.rust-lang.org/t/better-name-for-the-synthetic-types-generated-for-type-aliases/22590) for brainstorming alternative names.
